internal/action/render.go

```diff
@@ -0,0 +1,235 @@
+		// TODO(joelanford): Make a constant for above configs location label
+		cfg, err = declcfg.LoadDir(filepath.Join(tmpDir, configsDir))
+		if err != nil {
+			return nil, err
+		}
+	} else if _, ok := labels[bundle.PackageLabel]; ok {
+		img, err := registry.NewImageInput(ref, tmpDir)
+		if err != nil {
+			return nil, err
+		}
+
+		cfg, err = bundleToDeclcfg(img.Bundle)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		labelKeys := sets.StringKeySet(labels)
+		labelVals := []string{}
+		for _, k := range labelKeys.List() {
+			labelVals = append(labelVals, fmt.Sprintf("  %s=%s", k, labels[k]))
+		}
+		if len(labelVals) > 0 {
+			return nil, fmt.Errorf("render %q: image type could not be determined, found labels\n%s", ref, strings.Join(labelVals, "\n"))
+		} else {
+			return nil, fmt.Errorf("render %q: image type could not be determined: image has no labels", ref)
+		}
+	}
+	return cfg, nil
+}
+
+func sqliteToDeclcfg(ctx context.Context, dbFile string) (*declcfg.DeclarativeConfig, error) {
+	db, err := sqlite.Open(dbFile)
+	if err != nil {
+		return nil, err
+	}
+
+	migrator, err := sqlite.NewSQLLiteMigrator(db)
+	if err != nil {
+		return nil, err
+	}
+	if migrator == nil {
+		return nil, fmt.Errorf("failed to load migrator")
+	}
+
+	if err := migrator.Migrate(ctx); err != nil {
+		return nil, err
+	}
+
+	q := sqlite.NewSQLLiteQuerierFromDb(db)
+	m, err := sqlite.ToModel(ctx, q)
+	if err != nil {
+		return nil, err
+	}
+
+	cfg := declcfg.ConvertFromModel(m)
+	return &cfg, nil
+}
+
+func bundleToDeclcfg(bundle *registry.Bundle) (*declcfg.DeclarativeConfig, error) {
+	bundleProperties, err := registry.PropertiesFromBundle(bundle)
+	if err != nil {
+		return nil, fmt.Errorf("get properties for bundle %q: %v", bundle.Name, err)
+	}
+	relatedImages, err := getRelatedImages(bundle)
+	if err != nil {
+		return nil, fmt.Errorf("get related images for bundle %q: %v", bundle.Name, err)
+	}
+
+	dBundle := declcfg.Bundle{
+		Schema:        "olm.bundle",
+		Name:          bundle.Name,
+		Package:       bundle.Package,
+		Image:         bundle.BundleImage,
+		Properties:    bundleProperties,
+		RelatedImages: relatedImages,
+	}
+
+	return &declcfg.DeclarativeConfig{Bundles: []declcfg.Bundle{dBundle}}, nil
+}
+
+func getRelatedImages(b *registry.Bundle) ([]declcfg.RelatedImage, error) {
+	csv, err := b.ClusterServiceVersion()
+	if err != nil {
+		return nil, err
+	}
+
+	var objmap map[string]*json.RawMessage
+	if err = json.Unmarshal(csv.Spec, &objmap); err != nil {
+		return nil, err
+	}
+
+	rawValue, ok := objmap["relatedImages"]
+	if !ok || rawValue == nil {
+		return nil, err
+	}
+
+	var relatedImages []declcfg.RelatedImage
+	if err = json.Unmarshal(*rawValue, &relatedImages); err != nil {
+		return nil, err
+	}
+	return relatedImages, nil
+}
+
+func renderBundleObjects(cfg *declcfg.DeclarativeConfig) {
+	for bi, b := range cfg.Bundles {
+		props := b.Properties[:0]
+		for _, p := range b.Properties {
+			if p.Type != property.TypeBundleObject {
```

Comment:
Is this lossy? Will existing bundle objects contain valid data that `property.MustBuildBundleObjectData([]byte(obj)))` cannot rebuild?

Reply:
If I'm understanding the question, it should not be lossy. In the cases that we care about maintaining these properties (when existing indices are rendered), there are two places to be concerned about:
1. [Converting from sqlite image](https://github.com/operator-framework/operator-registry/blob/c95be3c757ffedf26e8ceb89b6d9462ced1604c6/pkg/api/api_to_model.go#L33).
2. [Rendering an existing DC index image/directory](https://github.com/operator-framework/operator-registry/blob/c95be3c757ffedf26e8ceb89b6d9462ced1604c6/internal/declcfg/load.go#L49-L55).

For bundle images, we intentionally do _not_ populate these because we the `olm.bundle.object` property is a temporary stopgap until the OLM cluster components no longer require the manifests to be sent in the GRPC response. Short term, index maintainers that require these properties would need to post-process the index to include these properties.